pkg/volume/csi/csi_attacher.go

```diff
@@ -351,18 +351,6 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 	}
 
 	err = saveVolumeData(dataDir, volDataFileName, data)
-	defer func() {
-		// Only if there was an error and volume operation was considered
-		// finished, we should remove the directory.
-		if err != nil && volumetypes.IsOperationFinishedError(err) {
-			// clean up metadata
-			klog.Error(log("attacher.MountDevice failed: %v", err))
-			if err := removeMountDir(c.plugin, deviceMountPath); err != nil {
-				klog.Error(log("attacher.MountDevice failed to remove mount dir after error [%s]: %v", deviceMountPath, err))
-			}
-		}
-	}()
-
 	if err != nil {
```

Comment:
Isn't a side-effect of this is, when `saveVolumeData` fails for whatever reason, the resulting path/file will not be removed? For example, the file got created but json encoding failed. Shouldn't the file be removed in that case? It isn't useful anyways.

Reply:
@gnufied I changed the `saveVolumeData ` function. If it fails to write data, the created file will be removed.